### DISCUSSION TITLE
Unpoly and server side redirect with hash fragment

### DISCUSSION THREAD

**Discussion by tomhillgreyridge** [unpoly/unpoly#453] [Q&A] [answered]
Does anyone know what the expected behaviour of Unpoly is if it does an up-submit to a server which returns a URL with a hash fragment?  i.e. https://myserver.com/longlist#item1234

To summarise the issue, I have an existing piece of code where there is a long list and every item in the list has an ID.  When you click an edit button on the item it opens a fresh page with a form on it.  Saving the form redirects back to the list page but with the id of the item as a hash fragment so that the browser scrolls back to show the updated item.

I'd like to do that with Unpoly so we are not reloading all CSS etc but it doesn't seem to do the scrolling part i.e. as soon as I add up-submit to the form, it stops following the hash fragment.

Is that expected behaviour?

**Comment by tomhillgreyridge** [comment]
For reference

- I know I can use up-save-scroll and up-scroll="reset" when dealing with edits, but this doesn't deal with adding new items (where I want to scroll to the new item but don't know it's ID until I have created it)
- I can't do "in place" editing as the list just shows an overview and the actual entity itself has quite a few fields.  Using Unpoly layers would be possible, but I'm currently just trying to replicate what we have currently just a bit smoother.

**Comment by triskweline** [comment, marked as answer]
JavaScript libraries like Unpoly cannot see the `#hash` of a redirect URL. This is a limitation of the browser APIs.

You may be able to work around this by also passing the hash as an [event](https://unpoly.com/X-Up-Events) that Unpoly *can* see. Something like that in the backend:

```ruby
def redirect_to_hash(url, hash)
  # Communicate the hash to client-side JavaScript
  response.headers['X-Up-Events'] = [ { type: 'scroll-to-hash', hash: hash }]

  # Regular redirect for full page loads
  redirect_to(url + "#" + hash)
end

redirect_to_hash '/companies', 'company-123'
```

And then pick up that event in the frontend:

```js
up.on('scroll-to-hash', ({ hash }) => up.viewport.revealHash(hash))
```

I haven't tested this code.

**Comment by tomhillgreyridge** [comment]
Thanks.  I was sure there would be a reason but wasn't sure what it was.  I'll test out your idea of a server side event and, if not, come up with a workaround
